File: backend-go/internal/application/validation.go
package application

import (
	"fmt"
	"strconv"
	"strings"
	"unicode/utf8"

	"github.com/shopspring/decimal"

	"github.com/pharaujo/finance/backend-go/internal/domain"
)

// Message templates copied from the System.ComponentModel.DataAnnotations
// resources, confirmed against bodies emitted by the running .NET API. Note
// that Required and EmailAddress name the field first while MinLength and
// MaxLength put "The field" first; that asymmetry is theirs, not a typo.
const (
	requiredMessage     = "The %s field is required."
	emailAddressMessage = "The %s field is not a valid e-mail address."
	minLengthMessage    = "The field %s must be a string or array type with a minimum length of '%d'."
	maxLengthMessage    = "The field %s must be a string or array type with a maximum length of '%d'."
)

// required mirrors RequiredAttribute with AllowEmptyStrings left at false: a
// string is missing when it is empty or entirely whitespace.
func required(errs *domain.ValidationError, field, value string) {
	if strings.TrimSpace(value) == "" {
		errs.Add(field, fmt.Sprintf(requiredMessage, field))
	}
}

// emailAddress mirrors EmailAddressAttribute, whose whole check is: exactly one
// '@', neither first nor last. It is applied to the raw value, before the
// service trims and lowercases it, because MVC validates the bound model.
func emailAddress(errs *domain.ValidationError, field, value string) {
	first := strings.IndexByte(value, '@')
	if first <= 0 || first == len(value)-1 || first != strings.LastIndexByte(value, '@') {
		errs.Add(field, fmt.Sprintf(emailAddressMessage, field))
	}
}

// minLength mirrors MinLengthAttribute. Lengths are counted in runes to match
// .NET's string.Length semantics closely enough for the ASCII inputs these
// fields hold; a UTF-16 surrogate pair would count as one here and two there.
func minLength(errs *domain.ValidationError, field, value string, limit int) {
	if utf8.RuneCountInString(value) < limit {
		errs.Add(field, fmt.Sprintf(minLengthMessage, field, limit))
	}
}

// maxLength mirrors MaxLengthAttribute.
func maxLength(errs *domain.ValidationError, field, value string, limit int) {
	if utf8.RuneCountInString(value) > limit {
		errs.Add(field, fmt.Sprintf(maxLengthMessage, field, limit))
	}
}

// The money bounds every [Range(typeof(decimal), ...)] in FinanceTracker's DTOs
// uses. They are strings because that is how the attribute spells them, and
// because the message quotes them verbatim.
const (
	// MoneyMinPositive is the lower bound of an amount that must be paid.
	MoneyMinPositive = "0.01"
	// MoneyMinZero is the lower bound of an amount that may be nothing.
	MoneyMinZero = "0.00"
	// MoneyMax is the largest value numeric(18,2) holds at two decimal places.
	MoneyMax = "999999999999.99"
)

const (
	rangeMessage = "The field %s must be between %s and %s."

	// missingMembersMessage stands in for the JsonException System.Text.Json
	// raises for an absent `required` member. MVC surfaces that under the "$"
	// key with the CLR type name in the text; the wording here is this API's,
	// the status (400) and the body shape are the .NET API's.
	missingMembersMessage = "The JSON payload was missing required properties, including the following: %s"
)

// decimalRange mirrors RangeAttribute over a decimal. Every money field in the
// DTOs shares MoneyMax as its upper bound, so only the lower one varies. An
// absent value is not its business: `required` members are checked by
// requiredMembers first.
func decimalRange(errs *domain.ValidationError, field string, value *domain.Money, minimum string) {
	if value == nil {
		return
	}

	low := decimal.RequireFromString(minimum)
	high := decimal.RequireFromString(MoneyMax)
	if value.LessThan(low) || value.GreaterThan(high) {
		errs.Add(field, fmt.Sprintf(rangeMessage, field, minimum, MoneyMax))
	}
}

// intRange mirrors RangeAttribute over an int, as used by the paging query.
func intRange(errs *domain.ValidationError, field string, value *int, minimum, maximum int) {
	if value == nil {
		return
	}
	if *value < minimum || *value > maximum {
		errs.Add(field, fmt.Sprintf(rangeMessage, field,
			strconv.Itoa(minimum), strconv.Itoa(maximum)))
	}
}

// requiredMembers reports the `required` members the payload left out. .NET
// fails deserialization for these, so nothing else about the body is checked;
// callers return as soon as this reports something.
func requiredMembers(errs *domain.ValidationError, missing []string) bool {
	if len(missing) == 0 {
		return false
	}
	errs.Add(domain.JSONBodyField, fmt.Sprintf(missingMembersMessage, strings.Join(missing, ", ")))
	return true
}

File: backend-py/app/domain/validation.py
"""The DataAnnotations rules the .NET API enforces, reproduced message for message.

The frontend renders these strings verbatim, and the other two backends emit
them byte-identically, so the wording below is fixed. Note the deliberate
asymmetry copied from .NET's resources: `required` and `emailAddress` put the
field name first ("The Email field ..."), while the length and range rules put
"The field" first ("The field Email ...").
"""

from __future__ import annotations

import re
from decimal import Decimal
from typing import Final

from app.core.errors import ValidationError
from app.domain.money import MONEY_MAX

# ModelState keys are the PascalCase property names of the .NET request DTOs,
# not the camelCase JSON names, because that is what the .NET API emits.
FIELD_EMAIL: Final = "Email"
FIELD_PASSWORD: Final = "Password"
FIELD_NAME: Final = "Name"
FIELD_CURRENCY: Final = "Currency"
FIELD_ICON: Final = "Icon"
FIELD_COLOR: Final = "Color"
FIELD_DESCRIPTION: Final = "Description"
FIELD_NOTES: Final = "Notes"
FIELD_AMOUNT: Final = "Amount"
FIELD_MONTH: Final = "Month"
FIELD_LIMIT: Final = "Limit"
FIELD_TARGET: Final = "TargetAmount"
FIELD_CURRENT: Final = "CurrentAmount"
FIELD_PAGE: Final = "Page"
FIELD_PAGE_SIZE: Final = "PageSize"
FIELD_SEARCH: Final = "Search"

# The key MVC's JSON reader uses for a body-level failure.
JSON_BODY_FIELD: Final = "$"

EMAIL_MAX_LENGTH: Final = 256
PASSWORD_MIN_LENGTH: Final = 8
PASSWORD_MAX_LENGTH: Final = 128
NAME_MAX_LENGTH: Final = 200
CURRENCY_MAX_LENGTH: Final = 8
ICON_MAX_LENGTH: Final = 64
COLOR_MAX_LENGTH: Final = 32
DESCRIPTION_MAX_LENGTH: Final = 500
NOTES_MAX_LENGTH: Final = 2000
SEARCH_MAX_LENGTH: Final = 200

REQUIRED_MESSAGE: Final = "The {field} field is required."
EMAIL_ADDRESS_MESSAGE: Final = "The {field} field is not a valid e-mail address."
MIN_LENGTH_MESSAGE: Final = (
    "The field {field} must be a string or array type with a minimum length of '{limit}'."
)
MAX_LENGTH_MESSAGE: Final = (
    "The field {field} must be a string or array type with a maximum length of '{limit}'."
)
RANGE_MESSAGE: Final = "The field {field} must be between {minimum} and {maximum}."
MISSING_MEMBERS_MESSAGE: Final = (
    "The JSON payload was missing required properties, including the following: {members}"
)
MONTH_FORMAT_MESSAGE: Final = "Month must be in YYYY-MM format."
INVALID_VALUE_MESSAGE: Final = "The value '{value}' is not valid for {field}."

# Looser than a real month parse on purpose: "2026-13" passes here and fails later.
MONTH_PATTERN: Final = re.compile(r"^\d{4}-\d{2}$")


def required(errs: ValidationError, field: str, value: str | None) -> None:
    """Whitespace-only counts as missing, as [Required] does."""
    if not (value or "").strip():
        errs.add(field, REQUIRED_MESSAGE.format(field=field))


def email_address(errs: ValidationError, field: str, value: str | None) -> None:
    """Exactly one '@', neither first nor last character, on the raw value."""
    text = value or ""
    if text.count("@") != 1 or text.startswith("@") or text.endswith("@"):
        errs.add(field, EMAIL_ADDRESS_MESSAGE.format(field=field))


def min_length(errs: ValidationError, field: str, value: str | None, limit: int) -> None:
    if len(value or "") < limit:
        errs.add(field, MIN_LENGTH_MESSAGE.format(field=field, limit=limit))


def max_length(errs: ValidationError, field: str, value: str | None, limit: int) -> None:
    if len(value or "") > limit:
        errs.add(field, MAX_LENGTH_MESSAGE.format(field=field, limit=limit))


def decimal_range(errs: ValidationError, field: str, value: Decimal | None, minimum: str) -> None:
    """A missing value is not this rule's business; the required-member check owns it."""
    if value is None:
        return
    if value < Decimal(minimum) or value > Decimal(MONEY_MAX):
        errs.add(field, RANGE_MESSAGE.format(field=field, minimum=minimum, maximum=MONEY_MAX))


def int_range(
    errs: ValidationError, field: str, value: int | None, minimum: int, maximum: int
) -> None:
    if value is None:
        return
    if value < minimum or value > maximum:
        errs.add(field, RANGE_MESSAGE.format(field=field, minimum=minimum, maximum=maximum))


def required_members(errs: ValidationError, missing: list[str]) -> bool:
    """Reports absent `required` members under "$".

    Returns True when something was missing, and every caller returns
    immediately on True — so no other rule runs. That short-circuit is the
    single most important ordering behaviour to keep: the .NET pipeline fails
    deserialisation before model validation ever starts.
    """
    if not missing:
        return False
    errs.add(JSON_BODY_FIELD, MISSING_MEMBERS_MESSAGE.format(members=", ".join(missing)))
    return True


def invalid_value(errs: ValidationError, value: str, field: str) -> None:
    errs.add(field, INVALID_VALUE_MESSAGE.format(value=value, field=field))


def raise_if_any(errs: ValidationError) -> None:
    if not errs.empty:
        raise errs
